pkg/controller/keycloakclient/keycloakclient_controller.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// Reconcile reads that state of the cluster for a KeycloakClient object and makes changes based on the state read
+// and what is in the KeycloakClient.Spec
+func (r *ReconcileKeycloakClient) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeycloakClient")
+
+	// Fetch the KeycloakClient instance
+	instance := &kc.KeycloakClient{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	err = r.validate(instance)
+	if err != nil {
+		return r.ManageError(instance, err)
+	}
+
+	r.adjustCR(instance)
+
+	keycloaks, err := r.getMatchingKeycloaks(instance)
+	if err != nil {
+		return r.ManageError(instance, err)
+	}
+
+	log.Info(fmt.Sprintf("found %v matching keycloak(s) for realm %v/%v", len(keycloaks.Items), instance.Namespace, instance.Name))
+
+	// The realm may be applicable to multiple keycloak instances,
+	// process all of them
+	for _, keycloak := range keycloaks.Items {
+		// Get an authenticated keycloak api client for the instance
+		authenticated, err := r.getAuthenticatedClient(keycloak)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+
+		// Compute the current state of the realm
+		log.Info(fmt.Sprintf("got authenticated client for keycloak at %v", keycloak.Status.InternalURL))
+		clientState := common.NewClientState(r.context, keycloak)
+
+		log.Info(fmt.Sprintf("read client state for keycloak %v/%v, realm %v/%v, client %v/%v",
+			keycloak.Namespace,
+			keycloak.Name,
+			instance.Namespace,
+			instance.Spec.RealmSelector.MatchExact,
+			keycloak.Namespace,
+			instance.Spec.Client.ClientID))
+
+		err = clientState.Read(r.context, instance, authenticated, r.client)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+
+		// Figure out the actions to keep the realms up to date with
+		// the desired state
+		reconciler := NewKeycloakClientReconciler(keycloak)
+		desiredState := reconciler.Reconcile(clientState, instance)
+		actionRunner := common.NewRealmActionRunner(r.context, r.client, r.scheme, instance, authenticated)
+
+		// Run all actions to keep the realms updated
+		err = actionRunner.RunAll(desiredState)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+	}
+
+	return reconcile.Result{Requeue: false}, r.manageSuccess(instance, instance.DeletionTimestamp != nil)
+}
+
+// Fills the CR with default values. Nils are not acceptable for Kubernetes.
+func (r *ReconcileKeycloakClient) adjustCR(cr *kc.KeycloakClient) {
```

Comment:
```suggestion
func (r *ReconcileKeycloakClient) setCrDefaults(cr *kc.KeycloakClient) {
```
Would prefer a more explicit name